File: CountDistinctSlices.py
def solution(M, A):
    # write your code in Python 2.7
    result = 0
    seen = [False] * (M+1)
    # two idxes
    front = back = 0
    while (front < len(A)):
        if (seen[A[front]] == False):
            # key point: how to calculate the accumulated number of different slices
            result += (front-back+1)
            if result > 1000000000:
                return 1000000000
            seen[A[front]] = True
            # remember to move idx at last
            front += 1
        else:
            while (A[back] != A[front]):
                seen[A[back]] = False
                back += 1
            seen[A[back]] = False
            back += 1
                
    return result

File: CountDistinctSlices.cpp
#include <algorithm>

int solution(int M, vector<int> &A) {
    // write your code in C++11
    int count = 0;
    vector<bool> used(M+1, false);
    int i = 0, j = 0;
    int n = A.size();
    while (j<n) {
        if (used[A[j]] == false) { // moving j
            used[A[j]] = true;
            count += j-i+1; // key: how to calculate the number of slices end at j and starting at i
            if (count > 1000000000) return 1000000000;
            j++;
        } else { // moving i
            used[A[i]] = false;
            i++;
        }       
    }
    return count;
}
